distributed/comm/ucx.py

```diff
@@ -0,0 +1,360 @@
+        The UCP endpoint.
+    address : str
+        The address, prefixed with `ucx://` to use.
+    deserialize : bool, default True
+        Whether to deserialize data in :meth:`distributed.protocol.loads`
+
+    Notes
+    -----
+    The read-write cycle uses the following pattern:
+
+    Each msg is serialized into a number of "data" frames. We prepend these
+    real frames with two additional frames
+
+        1. is_gpus: Boolean indicator for whether the frame should be
+           received into GPU memory. Packed in '?' format. Unpack with
+           ``<n_frames>?`` format.
+        2. frame_size : Unsigned int describing the size of frame (in bytes)
+           to receive. Packed in 'Q' format, so a length-0 frame is equivalent
+           to an unsized frame. Unpacked with ``<n_frames>Q``.
+
+    The expected read cycle is
+
+    1. Read the frame describing number of frames
+    2. Read the frame describing whether each data frame is gpu-bound
+    3. Read the frame describing whether each data frame is sized
+    4. Read all the data frames.
+    """
+
+    def __init__(
+        self, ep: ucp.Endpoint, address: str, listener_instance, deserialize=True
+    ):
+        logger.debug("UCX.__init__ %s %s", address, listener_instance)
+        self._ep = ep
+        assert address.startswith("ucx")
+        self.address = address
+        self.listener_instance = listener_instance
+        default_port = next(_PORT_COUNTER)
+        self._host, self._port = _parse_host_port(address, default_port)
+        self._local_addr = None
+        self.deserialize = deserialize
+        self.comm_flag = None
+
+        # finalizer?
+
+    @property
+    def local_address(self) -> str:
+        return self._local_addr
+
+    @property
+    def peer_address(self) -> str:
+        # XXX: This isn't quite for the server (from UCXListener).
+        # We need the port? Or the tag?
+        return self.address
+
+    async def write(self, msg: dict, serializers=None, on_error: str = "message"):
+        # msg can also be a list of dicts when sending batched messages
+        frames = await to_frames(
+            msg, serializers=serializers, on_error=on_error
+        )  # TODO: context=
+        gpu_frames = b"".join(
+            [
+                struct.pack("?", hasattr(frame, "__cuda_array_interface__"))
+                for frame in frames
+            ]
+        )
+        size_frames = b"".join([struct.pack("Q", nbytes(frame)) for frame in frames])
+
+        nframes = struct.pack("Q", len(frames))
+
+        meta = b"".join([nframes, gpu_frames, size_frames])
+
+        await self.ep.send_obj(meta)
+
+        for frame in frames:
+            await self.ep.send_obj(frame)
+        return sum(map(nbytes, frames))
+
+    async def read(self, deserializers=None):
+        resp = await self.ep.recv_future()
+        obj = ucp.get_obj_from_msg(resp)
+        nframes, = struct.unpack("Q", obj[:8])
+
+        gpu_frame_msg = obj[8 : 8 + nframes]
```

Comment:
What do you think about adding a small note on why an offset of `8` ?

Reply:
Sure.  Done in ecdf3cb